### ISSUE TITLE
Bug: [no-unnecessary-condition] False positive when value is modified in callback

### ISSUE THREAD

**Issue by VPanteleev-S7** [typescript-eslint/typescript-eslint#11231]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.2&fileType=.ts&code=KYDwDg9gTgLgBAMwK4DsDGMCWEVxsAZxgAoAoORVALjmIBMBzG4gSjgF4A%2BOANwkzpsuvfnVIsaKJAFsARsChwA3uTgAbYPAQRUdDogCGagsADcq5CmKsO3FRQrbd%2BmFCRnVAXxbmKmBLROKILKqhRQmkhQuACMvnCeqhEwUbgATOaJQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gdQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
export function test(
  fun: (dg: () => void) => void
): number {
  let found = false;
  fun(() => {
    found = true;
  });
  if (found) {
    return 1;
  }
  return 2;
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No diagnostics.

### Actual Result

```
Unnecessary conditional, value is always falsy. 8:7 - 8:12
```

### Additional Info

We don't know if `fun` will call `dg`, so the error seems incorrect.

**Comment by bradzacher** [comment]
This is mentioned in the docs:
https://typescript-eslint.io/rules/no-unnecessary-condition/#values-modified-within-function-calls
